plugin/src/main/java/org/opensearch/ml/action/connector/DeleteConnectorTransportAction.java

```diff
@@ -117,22 +118,41 @@ private void handleConnectorAccessValidationFailure(String connectorId, Exceptio
 
     private void checkForModelsUsingConnector(String connectorId, String tenantId, ActionListener<DeleteResponse> actionListener) {
         try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
-            SearchRequest searchRequest = new SearchRequest(ML_MODEL_INDEX);
             SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
             sourceBuilder.query(QueryBuilders.matchQuery(MLModel.CONNECTOR_ID_FIELD, connectorId));
             if (mlFeatureEnabledSetting.isMultiTenancyEnabled()) {
                 sourceBuilder.query(QueryBuilders.matchQuery(TENANT_ID, tenantId));
             }
-            searchRequest.source(sourceBuilder);
-            // TODO: Use SDK client not client.
-            client.search(searchRequest, ActionListener.runBefore(ActionListener.wrap(searchResponse -> {
-                SearchHit[] searchHits = searchResponse.getHits().getHits();
-                if (searchHits.length == 0) {
-                    deleteConnector(connectorId, actionListener);
-                } else {
-                    handleModelsUsingConnector(searchHits, connectorId, actionListener);
-                }
-            }, e -> handleSearchFailure(connectorId, e, actionListener)), context::restore));
+
+            SearchDataObjectRequest searchDataObjectRequest = new SearchDataObjectRequest.Builder()
+                .indices(ML_MODEL_INDEX)
+                .searchSourceBuilder(sourceBuilder)
+                .build();
+            sdkClient
+                .searchDataObjectAsync(searchDataObjectRequest, client.threadPool().executor(GENERAL_THREAD_POOL))
+                .whenComplete((sr, st) -> {
+                    context.restore();
+                    if (sr != null) {
+                        try {
+                            SearchResponse searchResponse = SearchResponse.fromXContent(sr.parser());
+                            SearchHit[] searchHits = searchResponse.getHits().getHits();
+                            if (searchHits.length == 0) {
+                                deleteConnector(connectorId, actionListener);
+                            } else {
+                                handleModelsUsingConnector(searchHits, connectorId, actionListener);
+                            }
+                        } catch (Exception e) {
+                            log.error("Failed to parse search response", e);
+                            actionListener
+                                .onFailure(
+                                    new OpenSearchStatusException("Failed to parse search response", RestStatus.INTERNAL_SERVER_ERROR)
+                                );
+                        }
+                    } else {
+                        Throwable cause = st.getCause() == null ? st : st.getCause();
```

Comment:
May be we should have a recursive method to identify the root cause as we know the exception can be wrapped with other exception. What do you think about this method:

```
private Throwable getRootCause(Throwable throwable) {
    Throwable cause = throwable;
    while (cause.getCause() != null && cause != cause.getCause()) {
        cause = cause.getCause();
    }
    return cause;
}
```

Reply:
See `unwrapCuase()` here.  This seems to do what you're asking for.  https://github.com/opensearch-project/OpenSearch/blob/802f2e6e4b21f27ddc6c01e7fc6f6cdcd69138d3/libs/core/src/main/java/org/opensearch/ExceptionsHelper.java#L127-L145

Reply:
Yes! Should we refactor the code accordingly then?

Reply:
Possibly?  But possibly not.   I intentionally only unwrapped one level.  Tracing the code, we know we're using `actionGet` so looking at the code we either throw the original exception (if it's a RTE) or it's a wrapped UncategorizedExecutionException containing the cause, which is what we're extracting here.    So that built in recursive method is a bit too deep.

While a util method is helpful here, a ternary one-liner is also pretty brief so I'm struggling to see the huge benefit of a refactor.  

So I'm open to it, but not thinking it's a high priority.